java/ql/lib/semmle/code/java/security/CleartextStorageAndroidFilesystemQuery.qll

```diff
@@ -0,0 +1,100 @@
+/**
+ * Provides classes and predicates to reason about cleartext storage in the Android filesystem
+ * (external or internal storage).
+ */
+
+import java
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.ExternalFlow
+import semmle.code.java.security.CleartextStorageQuery
+import semmle.code.java.security.Files
+import semmle.code.xml.AndroidManifest
+
+private class AndroidFilesystemCleartextStorageSink extends CleartextStorageSink {
+  AndroidFilesystemCleartextStorageSink() {
+    filesystemInput(_, this.asExpr()) and
+    // Make sure we are in an Android application.
+    exists(AndroidManifestXmlFile manifest)
+  }
+}
+
+/** The creation of an object that can be used to write to files to the local filesystem. */
+class LocalFileOpenCall extends Storable {
+  LocalFileOpenCall() {
+    this = any(DataFlow::Node sink | sinkNode(sink, "create-file")).asExpr().(Argument).getCall()
+  }
+
+  override Expr getAnInput() {
+    exists(FilesystemFlowConfig conf, DataFlow::Node n |
+      filesystemInput(n, result) and
+      conf.hasFlow(DataFlow::exprNode(this), n)
+    )
+  }
+
+  override Expr getAStore() {
+    exists(FilesystemFlowConfig conf, DataFlow::Node n |
+      filesystemStore(n, result) and
+      conf.hasFlow(DataFlow::exprNode(this), n)
+    )
+  }
+}
+
+/** Holds if `input` is written into `file`. */
+private predicate filesystemInput(DataFlow::Node file, Argument input) {
+  exists(DataFlow::Node write | sinkNode(write, "write-file") |
+    input = write.asExpr() or
+    isVarargs(input, write)
+  ) and
+  if input.getCall().getCallee().isStatic()
+  then file.asExpr() = input.getCall()
+  else file.asExpr() = input.getCall().getQualifier()
+}
+
+/** Holds if `arg` is part of `varargs`. */
+private predicate isVarargs(Argument arg, DataFlow::ImplicitVarargsArray varargs) {
+  arg.isVararg() and arg.getCall() = varargs.getCall()
+}
+
+/** Holds if `store` closes `file`. */
+private predicate filesystemStore(DataFlow::Node file, Call store) {
```

Comment:
Why is a filesystem store only a store if `file` is closed?
What about the `flush` method? Isn't it pretty much like the `close` method?

And is closing/flushing really necessary? Won't the data be written to the file anyway just by calling `write` on it? Or am I maybe missing something Android specific here?
(Assuming `java.io.Writer`)

Reply:
Good point. This is mainly to follow the `input-store` pattern that is used in the other cleartext storage queries. It's true that the file will be probably written even if `flush` or `close` aren't called, but I expect the files to be closed most of the time anyway (if not, the application has a leaking resource), which can be interpreted as a final "store everything that has been written" instruction.

I'll evaluate the query with public projects and see whether requiring the file/stream/writer to be closed causes many FNs. If that's the case, it could be fixed by considering `write` and similar methods both the input and the store for this query (and ignore `close` and `flush` altogether).